rs/execution_environment/src/canister_manager/tests.rs

```diff
@@ -6648,3 +6651,427 @@ fn memory_usage_updates_increment_subnet_available_memory() {
+            .unwrap();
+
+        env2.execute_round();
+        let wasm_result = env2.await_ingress(msg_id, MAX_TICKS).unwrap();
+
+        assert_matches!(wasm_result, WasmResult::Reply(r) if r == test_blob);
+    };
+    verify_stable_memory(canister_id2);
+
+    env2.stop_canister(canister_id2).unwrap();
+
+    let get_num_changes = |canister_id| {
+        env2.get_latest_state()
+            .canister_state(&canister_id)
+            .unwrap()
+            .system_state
+            .get_canister_history()
+            .get_total_num_changes()
+    };
+
+    let new_canister_id = CanisterId::from_u64(3 * CANISTER_IDS_PER_SUBNET - 1);
+    let new_version = 42;
+    let new_num_changes = 50;
+    let old_num_changes = get_num_changes(canister_id2);
+
+    let wasm_result = rename_canister(
+        &env1,
+        &env2,
+        canister_id1,
+        canister_id2,
+        new_canister_id,
+        2,
+        new_version,
+        new_num_changes,
+        false,
+    );
+    assert_matches!(wasm_result, WasmResult::Reply(r) if r == EmptyBlob.encode());
+
+    // Verify that the right canisters are present and can be run, and that version/canister history etc are as expected.
+    let verify_rename_happened = |old_canister_id,
+                                  new_canister_id,
+                                  rename_at_version,
+                                  current_version,
+                                  expected_num_changes,
+                                  expected_history_entry| {
+        let new_canister_exists = env2
+            .get_latest_state()
+            .canister_state(&new_canister_id)
+            .is_some();
+        let old_canister_exists = env2
+            .get_latest_state()
+            .canister_state(&old_canister_id)
+            .is_some();
+        assert!(new_canister_exists);
+        assert!(!old_canister_exists);
+
+        // The version should have been updated
+        assert_eq!(
+            current_version,
+            env2.get_latest_state()
+                .canister_state(&new_canister_id)
+                .unwrap()
+                .system_state
+                .canister_version
+        );
+        assert_eq!(
+            expected_num_changes,
+            env2.get_latest_state()
+                .canister_state(&new_canister_id)
+                .unwrap()
+                .system_state
+                .get_canister_history()
+                .get_total_num_changes()
+        );
+        let history_entry = env2
+            .get_latest_state()
+            .canister_state(&new_canister_id)
+            .unwrap()
+            .system_state
+            .get_canister_history()
+            .get_changes(1)
+            .next()
+            .unwrap()
+            .clone();
+
+        assert_eq!(history_entry.canister_version(), rename_at_version);
+        assert_eq!(history_entry.details(), &expected_history_entry);
+
+        env2.start_canister(new_canister_id).unwrap();
```

Comment:
Maybe we could move the call to start the canister to the outside (starting the canister is not really part of renaming "verification").